lib/AST/Decl.cpp

```diff
@@ -2699,6 +2699,20 @@ bool NominalTypeDecl::isOptionalDecl() const {
   return this == getASTContext().getOptionalDecl();
 }
 
+ArrayRef<VarDecl *>
+NominalTypeDecl::getAllTFParameters() {
+  if (TFParameters) return *TFParameters;
+  SmallVector<VarDecl *, 2> parameters;
+  for (auto member : getMembers()) {
+    auto varDecl = dyn_cast<VarDecl>(member);
+    if (!varDecl) continue;
+    if (varDecl->getAttrs().hasAttribute<TFParameterAttr>())
+      parameters.push_back(varDecl);
+  }
+  TFParameters = parameters;
```

Comment:
This involves a copy. I suggest the following:

```cpp
if (TFParameters) return *TFParameters;
TFParameters = SmallVector<VarDecl *, 2>();
for (auto member : getMembers()) {
  auto varDecl = dyn_cast<VarDecl>(member);
  if (!varDecl) continue;
  if (varDecl->getAttrs().hasAttribute<TFParameterAttr>())
    TFParameters->push_back(varDecl);
}
return *TFParameters;
```

Reply:
Thanks! Done in https://github.com/apple/swift/pull/18171/commits/99b20ffce93152d2e7d2b19d8e2371f1a601b1d3.